### PR TITLE
api: gpio: Clean up GPIO dt-bindings flags

### LINKED ISSUE #10339: gpio: Cleanup flags
This issue is to discuss reworking the current GPIO flags, today we have:

```
/* direction */
#define GPIO_DIR_IN             (0 << 0)
#define GPIO_DIR_OUT            (1 << 0)

/* interrupt */
#define GPIO_INT                (1 << 1)
#define GPIO_INT_ACTIVE_LOW     (0 << 2)
#define GPIO_INT_ACTIVE_HIGH    (1 << 2)
#define GPIO_INT_CLOCK_SYNC     (1 << 3)
#define GPIO_INT_DEBOUNCE       (1 << 4)
#define GPIO_INT_LEVEL          (0 << 5)
#define GPIO_INT_EDGE           (1 << 5)
#define GPIO_INT_DOUBLE_EDGE    (1 << 6)

/* polarity */
#define GPIO_POL_NORMAL         (0 << GPIO_POL_POS)
#define GPIO_POL_INV            (1 << GPIO_POL_POS)

/* PULL up/down */
#define GPIO_PUD_NORMAL         (0 << GPIO_PUD_POS)
#define GPIO_PUD_PULL_UP        (1 << GPIO_PUD_POS)
#define GPIO_PUD_MASK           (3 << GPIO_PUD_POS)

#define GPIO_DS_LOW_MASK (0x3 << GPIO_DS_LOW_POS)
#define GPIO_DS_DFLT_LOW (0x0 << GPIO_DS_LOW_POS)
#define GPIO_DS_ALT_LOW (0x1 << GPIO_DS_LOW_POS)
#define GPIO_DS_DISCONNECT_LOW (0x3 << GPIO_DS_LOW_POS)
#define GPIO_DS_DFLT_HIGH (0x0 << GPIO_DS_HIGH_POS)
#define GPIO_DS_ALT_HIGH (0x1 << GPIO_DS_HIGH_POS)
#define GPIO_DS_DISCONNECT_HIGH (0x3 << GPIO_DS_HIGH_POS)
```

### PR DESCRIPTION
**Pull request by mnkp**
This PR attempts to resolve the concerns and discussions of issue #10339 and is based on work in #11880 by @pabigot.

The main differences to #11880 are as follows:
- multi-bit flags are defined using following standard pattern
```
#define GPIO_MULTI_BIT_FIELD             (N << GPIO_MULTI_BIT_FIELD_SHIFT)
```
where 'N' is the value/state we want to encode.
- no _SHIFT and _MASK macros for single bit flags.
- GPIO_POL_ flags are deprecated as their function overlaps with the new `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags.
- the documentation of `gpio_pin_read()`, `gpio_pin_write()` functions was updated to clarify usage of the new GPIO_ACTIVE_ flags.

The following design goals were applied:
- only issues discussed in #10339 are addressed. Otherwise the changes are kept to minimum. Especially this PR does not attempt to align the API with the future pinctrl API (e.g. #6760).
- to be able to keep DTS of GPIO nodes close to the [Linux's dt-bindings gpio.h](https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/gpio.h)
Linux implementation is followed.

One extra point which we likely should revisit are [issues](https://github.com/zephyrproject-rtos/zephyr/issues/10339#issuecomment-426442529) raised by @DWagenk. Current implementation of pin read/write and interrupt API is not consistent.

I have not yet verified that generated documentation is formatted correctly.

Fixes #10339

### CODE REVIEW

**Comment by mnkp** [comment]
In the next few days I'll try to prepare another PR with more involved changes to GPIO flags where I will try to align the design to the future pinctrl API as proposed in #6760 by @b0661. In general the pinctrl API gives an impression of a better thought of than the GPIO one.

**Review by pabigot — COMMENTED** [review]
There are a couple capabilities in #11880 that don't seem to be supported by this version:

* The ability to configure direction as simultaneously input and output, a feature supported by multiple drivers including SX1509B and Nordic.  By using separate bits we get better control of power consumption by explicitly disabling input and output buffers.
* The ability to explicitly configure drive strength as low or high, where the defaults differ between SX1509B and Nordic (noted as lacking)
* The ability to specify the initial level for output GPIOs.

There may be others.  It also lacks an example of how to implement the extensions it provides, and it doesn't build.

All that can of course be addressed, but fundamentally I don't see anything here that justifies spending more time on a solution that is not demonstrably superior to the one that was proposed in early December and already has four approvals plus a LGTM.

**Comment by mnkp** [comment]
> * The ability to configure direction as simultaneously input and output, a feature supported by multiple drivers including SX1509B and Nordic.  By using separate bits we get better control of power consumption by explicitly disabling input and output buffers.

The functionality was not changed. Pin configured as GPIO_DIR_OUT works simultaneously as input. Current proposal doesn't have ability to disable pin. This could be addressed by adding GPIO_DIR_DISABLED flag. But a better approach would be to use pinctrl like interface which provides a more sophisticated approach to configuring pin drive capabilities.

> * The ability to explicitly configure drive strength as low or high, where the defaults differ between SX1509B and Nordic (noted as lacking)

It's mentioned on the To Do list, will address shortly.

> * The ability to specify the initial level for output GPIOs.

Supported by GPIO_DIR_OUT_LOW, GPIO_DIR_OUT_HIGH flags

File: include/dt-bindings/gpio/gpio.h
Comment (pabigot) [inline include/dt-bindings/gpio/gpio.h]: If part of your motivation for rejecting #11880 was objection to defining offset macros for single-bit fields, you really ought to remove this one.

**Comment by pabigot** [comment]
> > * The ability to configure direction as simultaneously input and output, a feature supported by multiple drivers including SX1509B and Nordic.
> 
> The functionality was not changed.

Yes, which is my point.  Ability to control input enable explicitly was added in #11880, and is relevant to at least two existing GPIO driver implementations.

> But a better approach would be to use pinctrl like interface which provides a more sophisticated approach to configuring pin drive capabilities.

That's an opinion that can be defended or challenged.  But we don't have pinctrl.  We won't have pinctrl for 1.14-LTS.  I would like to have the functionality now.

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [reply include/dt-bindings/gpio/gpio.h]: All GPIO_POL_ flags were deprecated and are going to be removed in the future. `GPIO_POL_SHIFT` is required to define `GPIO_POL_MASK` which is used in the code base.

**Comment by mnkp** [comment]
> > > * The ability to configure direction as simultaneously input and output, a feature supported by multiple drivers including SX1509B and Nordic.
> > 
> > 
> > The functionality was not changed.
> 
> Yes, which is my point. Ability to control input enable explicitly was added in #11880, and is relevant to at least two existing GPIO driver implementations.

It was always possible "to configure direction as simultaneously input and output". I've added also `GPIO_DIR_DISABLED` flag to disable / disconnect the pin. What kind of use case are you concerned about?

File: include/dt-bindings/gpio/gpio.h
Comment (pabigot) [reply include/dt-bindings/gpio/gpio.h]: So there is existing precedent in Zephyr for providing mask and shift defines for single-bit fields.  Good to know.

**Review by b0661 — COMMENTED** [review]
LGTM

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [reply include/dt-bindings/gpio/gpio.h]: `GPIO_DIR_IN` configures pin as input only. It's actually `GPIO_DIR_OUT` that configures pin as an output but allows simultaneously to read values from the pin. This is also the current behavior. I can add a relevant note to the `GPIO_DIR_OUT` macro?

**Comment by carlescufi** [comment]
@Mierunski could you please take a look at this PR in the context of #11880?

**Comment by pabigot** [comment]
> * GPIO_POL_ flags are deprecated as their function overlaps with the new `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags.

Wait, what?  No they don't.  At least I don't think so.  I thought they were meant to say "configure the device to invert the signal polarity".  (Whatever that means.)

At least that's how they're used in most drivers now.

**Comment by pabigot** [comment]
> * the documentation of `gpio_pin_read()`, `gpio_pin_write()` functions was updated to clarify usage of the new GPIO_ACTIVE_ flags.

That's not right either.  The way I've always seen the `GPIO_ACTIVE_` flags used is as an indicator of whether the active state of the signal is high or low.  E.g. a pressed button may result in a connection to ground (active low) or to a current source (active high).  Similarly for LEDs where you write a zero to illuminate an active low LED and a one to illuminate an active high LED.

They certainly don't affect the value written by `gpio_pin_write`, and they had better not affect the value returned from `gpio_pin_read`.

File: include/gpio.h
Comment (pabigot) [inline include/gpio.h]: This isn't what those flags are intended to convey, nor does this describe how this function actually behaves (unless I am *very* confused...).

File: include/gpio.h
Comment (pabigot) [inline include/gpio.h]: The function does not adjust its return value based on `GPIO_ACTIVE_LOW`.

File: include/dt-bindings/gpio/gpio.h
Comment (pabigot) [inline include/dt-bindings/gpio/gpio.h]: These two are not equivalent.  In the intended use as documented in #11880 `GPIO_ACTIVE_foo` is informational, indicating the active level of the signal.  It's specifically needed in the DTS bindings to indicate whether an LED (for example) needs zero or one written to turn it on.

Polarity inversion is a configuration that affects how the driver sets up the GPIO peripheral.

File: include/dt-bindings/gpio/gpio.h
Comment (pabigot) [inline include/dt-bindings/gpio/gpio.h]: Same: inverted polarity does not mean active low.

File: include/dt-bindings/gpio/gpio.h
Comment (pabigot) [inline include/dt-bindings/gpio/gpio.h]: This doesn't seem right.  In #11880 the base conditions are:
* `GPIO_IO_IN_ENABLED` enables or disables input from the GPIO.  On some devices this corresponds to connecting or disconnecting the input buffer.
* `GPIO_IO_OUT_ENABLED` enables or disables output from the GPIO.

I think the equivalent here would be:
* neither input nor output: `GPIO_DIR_DISABLED`
* input-only: `GPIO_DIR_IN` (== 0)
* output-only: not expressible?
* input and output: `GPIO_DIR_OUT | GPIO_DIR_IN` == `GPIO_DIR_OUT`.

So this fails to provide desirable functionality.

**Review by pabigot — DISMISSED** [review]
Some more issues resulting from a closer look.

**Comment by dleach02** [comment]
For all deprecated defines I would suggest compressing the defines and having a single comment block around them that states they will be deprecated. Don't attempt to document them any further because only existing implementation will have them in use and new implementations you want to discourage usage.

**Review by galak — CHANGES_REQUESTED** [review]
Is there any particular reason to have how bit positions 0, 1, and 2 are defined differ from the linux dt-binding gpio.h:

```
/* Bit 0 express polarity */
#define GPIO_ACTIVE_HIGH 0
#define GPIO_ACTIVE_LOW 1

/* Bit 1 express single-endedness */
#define GPIO_PUSH_PULL 0
#define GPIO_SINGLE_ENDED 2

/* Bit 2 express Open drain or open source */
#define GPIO_LINE_OPEN_SOURCE 0
#define GPIO_LINE_OPEN_DRAIN 4

/*
 * Open Drain/Collector is the combination of single-ended open drain interface.
 * Open Source/Emitter is the combination of single-ended open source interface.
 */
#define GPIO_OPEN_DRAIN (GPIO_SINGLE_ENDED | GPIO_LINE_OPEN_DRAIN)
#define GPIO_OPEN_SOURCE (GPIO_SINGLE_ENDED | GPIO_LINE_OPEN_SOURCE)

```

**Comment by mnkp** [comment]
> > * the documentation of `gpio_pin_read()`, `gpio_pin_write()` functions was updated to clarify usage of the new GPIO_ACTIVE_ flags.
> 
> That's not right either. The way I've always seen the `GPIO_ACTIVE_` flags used is as an indicator of whether the active state of the signal is high or low. E.g. a pressed button may result in a connection to ground (active low) or to a current source (active high). Similarly for LEDs where you write a zero to illuminate an active low LED and a one to illuminate an active high LED.

>  In the intended use as documented in #11880 GPIO_ACTIVE_foo is informational, indicating the active level of the signal. It's specifically needed in the DTS bindings to indicate whether an LED (for example) needs zero or one written to turn it on.

That was my impression when reviewing #11880 that the usage of `GPIO_ACTIVE_` flags was not understood and clarified well enough. As per [comment](https://github.com/zephyrproject-rtos/zephyr/issues/10339#issuecomment-426359428) from @galak the GPIO_POL_ flags should be removed and replaced by `GPIO_ACTIVE_`.

`GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` are used by the [Linux's dt-bindings gpio.h](https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/gpio.h). And we want to keep Zephyr's usage of GPIO DTS node close to the Linux equivalent. `GPIO_ACTIVE_` flags are not merely informational. Their meaning is exactly as documented in the description of `gpio_pin_read()`, `gpio_pin_write()` functions.

Linux documents its usage of [Active-High and Active-Low](https://www.kernel.org/doc/html/v4.17/driver-api/gpio/intro.html#active-high-and-active-low) flags. This is also mentioned in the description of gpio [get_value](https://www.kernel.org/doc/html/v4.17/driver-api/gpio/index.html#c.gpiod_get_value), set_value functions.

**Comment by pabigot** [comment]
I don't read that comment from @galak quite the same way, though I can see that interpretation.

The usage of active high/low flags in Linux is exactly as I've described; the difference is that Linux apparently uses them to control the sense of read/write operations in their API.  Zephyr's read/write GPIO functions affect the signal value directly, not the logical value as Linux does.  Maybe that was not the intent, but AFAICT all existing use of `GPIO_INT_ACTIVE_foo` is restricted to detecting the signal level when the interrupt is asserted.  That's not something we can change now.

The `GPIO_POL_` flags also have an existing behavior that we can't suddenly change.

**Comment by galak** [comment]
> I don't read that comment from @galak quite the same way, though I can see that interpretation.
> 
> The usage of active high/low flags in Linux is exactly as I've described; the difference is that Linux apparently uses them to control the sense of read/write operations in their API. Zephyr's read/write GPIO functions affect the signal value directly, not the logical value as Linux does. Maybe that was not the intent, but AFAICT all existing use of `GPIO_INT_ACTIVE_foo` is restricted to detecting the signal level when the interrupt is asserted. That's not something we can change now.
> 
> The `GPIO_POL_` flags also have an existing behavior that we can't suddenly change.

My comment was about the bit position and names.  There's a subtle point here about Linux compatibility and Device Tree.  We should be aiming to utilize and conform to the Device Tree specs that are utilized in Linux, that's the canonical device tree definition repository.  If there are some area's that the DTS binding aren't clear or need clarity we should aim to get that clarity.

**Comment by pabigot** [comment]
Discussion in dev review telecon 2019-01-24 leaves this open until next week and more review.

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [reply include/dt-bindings/gpio/gpio.h]: >     * output-only: not expressible?

This is indeed not supported. There is only so much one can do by extending GPIO_DIR_ flags. To have it added we would need to deprecate GPIO_DIR_ flags and provide new configuration based on pinctrl interface which does it in a clean and consistent way.

Comment (pabigot) [reply include/dt-bindings/gpio/gpio.h]: I thought I managed to do it.  Though that was only because existing code uses:
```
               if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
```
taking advantage of the power of masks for single-bit fields....

**Comment by mnkp** [comment]
I have reworked the way we configure input / output mode. Rather than using `GPIO_DIR_*` flags which are now deprecated I'm using a new `GPIO_INPUT_*`, `GPIO_OUTPUT_*` flags. This approach is consistent with [pinctrl-bindings.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-bindings.txt) as well as [bindings gpio.txt](https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/gpio.txt) used by Linux. What it means is that if we want to be compliant with Linux DTS for gpio and pinctlr nodes this API will support it.

I've added also usage examples of `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags. As well as modified two drivers: gpio_sam.c, gpio_gecko.c to showcase the new GPIO API.

As per request from @galak the encoding of flags used in DTS is now identical as in [linux dt-binding gpio.h](https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/gpio.h).

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [reply include/dt-bindings/gpio/gpio.h]: I've reworked the flags to allow independently configure input/output mode. This is consistent with the approach used by Linux pinctr/gpio drivers. The output only mode is now possible.

File: include/dt-bindings/gpio/gpio.h
Comment (mnkp) [reply include/dt-bindings/gpio/gpio.h]: I've reworked the way input/output mode is configured. `GPIO_DIR_OUT` flag is now deprecated.

**Comment by mnkp** [comment]
> The usage of active high/low flags in Linux is exactly as I've described [in #11880]; the difference is that Linux apparently uses them to control the sense of read/write operations in their API.

This statement doesn't quite make sense. Either the implementation of `GPIO_ACTIVE_LOW` flag in #11880 is consistent with the way it's used by Linux driver API and in Linux DTS or it isn't. That's an important difference you're trying to overlook.

> The `GPIO_POL_` flags also have an existing behavior that we can't suddenly change.

The very purpose of this PR is to rework usage of GPIO flags. So the statement once again doesn't quite make sense. Double so, taking into account that the usage of `GPIO_POL_` flags wasn't very well defined from the very beginning and #11880 changed their definition in a way that makes them even more opaque.

**Comment by pabigot** [comment]
> > The usage of active high/low flags in Linux is exactly as I've described [in #11880]; the difference is that Linux apparently uses them to control the sense of read/write operations in their API.
> 
> This statement doesn't quite make sense. Either the implementation of `GPIO_ACTIVE_LOW` flag in #11880 is consistent with the way it's used by Linux driver API and in Linux DTS or it isn't. That's an important difference you're trying to overlook.

I don't think I'm overlooking it.  I see it as two things: what do the flags indicate, and how do the driver methods work.

The flag semantics I described in #11880 is generally consistent with the first half of [the reference you gave](https://www.kernel.org/doc/html/v4.17/driver-api/gpio/intro.html#active-high-and-active-low): they describe what line level indicates "active" for the signal.

I agree that the last half (how the flags affect interaction with the signal) is different, but that's not my fault.  In #11880 I chose to leave the behavior of `gpio_pin_write` and `gpio_pin_read` the way it has always been in Zephyr (reads and writes the electrical signal level) rather than the way it's done in Linux (reads and writes the logical signal level).

I would in principal be open to changing the driver functions to work with the logical levels, but we'd have to change a few things:
* The default (zero) value must be allocated to `GPIO_ACTIVE_HIGH` so that writing a 1 has the same meaning it does now.  This is what you've done, and now that I think about it it makes sense, so I'm going to update #11880 to match.
* Every driver would have to be updated to store the configured active level and use it to mutate the write operand and read result.  (Unless they happen to have an invert capability that provides those semantics.)
* There should probably be a new `gpio_pin_configuration()` method that can be used to query the pin configuration so somebody who wants the electrical rather than signal level can reconstruct it.

That seems a bit much to change, but if there's consensus that it's the right way forward and every GPIO driver maintainer commits to updating their code within the next few weeks (so it's available for several weeks of pre-release testing) I'd be willing to change #11880.

> > The `GPIO_POL_` flags also have an existing behavior that we can't suddenly change.
> 
> The very purpose of this PR is to rework usage of GPIO flags. So the statement once again doesn't quite make sense. Double so, taking into account that the usage of `GPIO_POL_` flags wasn't very well defined from the very beginning and #11880 changed their definition in a way that makes them even more opaque.

I'm not aware of doing anything to change the `GPIO_POL_` flags.  Drivers that respect them have exactly the same effect as they always did.  I don't see a motivation for deprecating them, and somebody might well be depending on the existing behavior so I really don't think we can change it.  It's certainly driver-dependent, but again---somebody put it into the API, other people used it to control configurations, and we need to be careful about taking things away once that's happened.

**Comment by mnkp** [comment]
Since during the API call there seem to be misunderstanding about the purpose of `GPIO_ACTIVE_LOW`, `GPIO_ACTIVE_HIGH` flags I would like to further clarify their usage. In principle they allow application to configure pin as working with positive or negative logic. Here is a short, random, on-line article that clarifies the concept: [positive vs. negative logic](http://electriciantraining.tpub.com/14185/css/Positive-And-Negative-Logic-85.htm). Linux documentation of gpio driver also provides abundant information.

If an application prefers to control CS (Chip Select) pin which is typically active low by writing `0` as a pin value it means that the pin should be configured as `GPIO_ACTIVE_HIGH` since we don't want to invert the logic. However, forcing a GPIO driver to work exclusively with positive logic is impractical. There are many use cases in embedded world where the only way to write simple, portable code is to have this functionality implemented in the driver. The typical examples are controlling the LEDs or power switches. They can be activated by either low or high output state depending on the board design or even the board version. To control a power switch application should not care about the specifics of the hardware but simply write `1` to enable the switch and `0` to disable it.

**Comment by pabigot** [comment]
Please see https://github.com/zephyrproject-rtos/zephyr/pull/11880#issuecomment-459269558 for a proposed compromise that supports existing usage while clarifying existing API and adding API to support positive/negative logic.

File: boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts
Comment (anangl) [inline boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts]: Wasn't the `GPIO_ACTIVE_LOW` flag supposed to be used also to specify active state of inputs. One may not want to use interrupts to handle buttons. Sometimes just reading the current state of the button could be desired.

Comment (mnkp) [reply boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts]: That's correct. I will change this line to
```
	gpios = <&gpiof 6 (GPIO_ACTIVE_LOW | GPIO_INT_LOW)>;
```

File: include/gpio.h
Comment (anangl) [reply include/gpio.h]: I am against changing the definition of how `gpio_pin_write` and `gpio_pin_read` are expected to behave. This would require alteration of all existing implementations of GPIO drivers (as these flags cannot be just not supported).
In my opinion the `GPIO_ACTIVE_*` flags should be handled outside of particular GPIO drivers, for example in wrappers as I proposed in https://github.com/zephyrproject-rtos/zephyr/pull/11880#issuecomment-461011334.

Comment (anangl) [inline include/gpio.h]: I am against changing the definition of how `gpio_pin_write` and `gpio_pin_read` are expected to behave. This would require alteration of all existing implementations of GPIO drivers (as the `GPIO_ACTIVE_LOW` flag cannot be just not supported).
In my opinion the `GPIO_ACTIVE_*` flags should be handled outside of particular GPIO drivers, for example in wrappers as I proposed in https://github.com/zephyrproject-rtos/zephyr/pull/11880#issuecomment-461011334.

Comment (mnkp) [reply include/gpio.h]: I provided a bit more elaborate answer in my [comment](https://github.com/zephyrproject-rtos/zephyr/pull/11880#issuecomment-461080940) in #11880. In short `GPIO_ACTIVE_LOW` is just a different name for `GPIO_POL_INV` flag which controls the value returned by `gpio_pin_write` and `gpio_pin_read` functions. Unfortunately, the documentation of these functions did not state it explicitly. The only reason we rename the `GPIO_POL_INV` flag to `GPIO_ACTIVE_LOW` is to be compatible with the definition of DTS gpio node in Linux.

**Comment by mnkp** [comment]
Superseded by #16648.

**Pull request closed without merging**
